File: tools/build_employee_import.js
/**
 * Turn the Factor HR employee master into an ERPNext Data Import file.
 *
 *     node tools/build_employee_import.js                 # active staff only
 *     node tools/build_employee_import.js --include-left  # everyone, leavers too
 *
 * Writes `data/out/employee-import.csv`, which is gitignored along with
 * everything else under `data/`.
 *
 * Two things this deliberately does **not** do:
 *
 *   - **It never guesses a company.** Six spellings in Factor HR map onto five
 *     ERPNext companies through the table below, and a row whose company is not
 *     in that table is rejected rather than defaulted. A guessed company is a
 *     person in the wrong payroll, discovered at month end.
 *   - **It does not set `reports_to`.** A manager has to exist before anyone can
 *     report to them, so that is a second pass — see backfill_employee_links.js.
 */

import path from "node:path";
import { parseArgs } from "node:util";
import ExcelJS from "exceljs";
import { ensureDir, readFirstSheet, writeCsv } from "./lib/sheets.js";
import { col, columnsOf, findHeader, get } from "./lib/factohr.js";

const SOURCE = "data/factohr/Employee Detail Report.xlsx";
const OUT_DIR = "data/out";

/* Factor HR spelling -> ERPNext Company. Confirmed with Manna on 22 Aug 2026.
 *
 * Note the two rows that collapse onto Manna Treads: HI-TECH PRETREADS is the
 * same company, not a separate one, so its 112 people join the 4 already there. */
const COMPANY = {
	"MANNA RUBBER PRODUCTS PVT.LTD.": "Manna Rubber Products Private Limited",
	"HI-TECH PRETREADS": "Manna Treads",
	"MANNA TREADS PVT.LTD": "Manna Treads",
	"HI-TECH RUBBER INDUSTRIES": "Hi-Tech Rubber Industries",
	"MANNA TYRE RETREADS": "Manna Tyre Retreads",
	"MANNA GROUP H-QTRS": "Manna Group Headquarters",
};

/* ERPNext names a Department "<name> - <company abbr>", so a bare "Production"
   links to nothing and the import fails a row at a time. Designation is not
   suffixed. This caught us on the first build. */
const ABBR = {
	"Manna Rubber Products Private Limited": "MRPPL",
	"Manna Treads": "MT",
	"Manna Tyre Retreads": "MTR",
	"Hi-Tech Rubber Industries": "HRI",
	"Manna Group Headquarters": "MGHQ",
	"Manna Tyre UAE": "MRU",
};

// ERPNext's Employee.status is a Select with exactly these options.
const STATUS = { active: "Active", inactive: "Left" };

const GENDER = { male: "Male", female: "Female" };

// Columns written, in the order ERPNext's importer likes to see them.
const HEADERS = [
	"ID",
	"Employee Name",
	"First Name",
	"Last Name",
	"Gender",
	"Date of Birth",
	"Date of Joining",
	"Relieving Date",
	"Status",
	"Company",
	"Department",
	"Designation",
	"Employment Type",
	"Attendance Device ID (Biometric/RF tag ID)",
	"Cell Number",
	"Factor HR ID",
];

const MONTHS = {
	jan: "01", feb: "02", mar: "03", apr: "04", may: "05", jun: "06",
	jul: "07", aug: "08", sep: "09", oct: "10", nov: "11", dec: "12",
};

/** Factor HR emits '1966-12-04 00:00:00'. ERPNext wants '1966-12-04'. */
function asDate(value) {
	if (!value) return "";
	let m = /^(\d{4}-\d{2}-\d{2})/.exec(value);
	if (m) return m[1];
	m = /^(\d{2})-([A-Za-z]{3})-(\d{4})/.exec(value);
	if (m) {
		const mon = MONTHS[m[2].toLowerCase()];
		if (mon) return `${m[3]}-${mon}-${m[1]}`;
	}
	return "";
}

/**
 * ERPNext requires `first_name`; `last_name` is optional.
 *
 * Many rows here are a single word, and several are a full name in one field
 * with no reliable split. Putting everything in `first_name` and letting
 * `employee_name` carry the display form is honest — inventing a surname from
 * the last token would corrupt names that do not work that way.
 */
function splitName(full) {
	const s = full.split(/\s+/).filter(Boolean);
	if (!s.length) return ["", ""];
	if (s.length === 1) return [s[0], ""];
	return [s.slice(0, -1).join(" "), s[s.length - 1]];
}

const capitalize = (w) => (w ? w[0].toUpperCase() + w.slice(1).toLowerCase() : w);

/** Factor HR shouts everything. Title-case it, but leave initialisms alone. */
function tidy(value) {
	const words = String(value || "").split(/\s+/).filter(Boolean);
	if (!words.length) return "";
	return words.map((word) => {
		/* Q.C., H.R, PVT — anything with a dot, or 3 letters or fewer that is
		   all caps, is almost certainly an initialism rather than a shouted
		   word. */
		const allCaps = word === word.toUpperCase() && /[A-Z]/.test(word);
		if (word.includes(".") || (allCaps && word.length <= 3)) return word;
		return capitalize(word);
	}).join(" ");
}

const departmentName = (raw, company) => (raw ? `${tidy(raw)} - ${ABBR[company]}` : "");

const padEnd = (s, n) => String(s).padEnd(n);

async function main() {
	const { values } = parseArgs({
		options: {
			source: { type: "string", default: SOURCE },
			"include-left": { type: "boolean", default: false },
		},
	});

	const rows = await readFirstSheet(values.source, { maxRows: 3000 });
	const index = findHeader(rows, "Emp Code");
	if (index === null) {
		process.stderr.write(`Could not find an 'Emp Code' column in ${values.source}\n`);
		process.exit(1);
	}
	const headers = columnsOf(rows[index]);

	const ix = {
		code: col(headers, "Emp Code"),
		name: col(headers, "Full Name"),
		company: col(headers, "Company Name"),
		status: col(headers, "Status"),
		machine: col(headers, "Machine Code"),
		dob: col(headers, "Birth Date"),
		doj: col(headers, "Joining Date"),
		left: col(headers, "Leaving Date"),
		gender: col(headers, "Gender"),
		dept: col(headers, "Department"),
		desig: col(headers, "Designation"),
		etype: col(headers, "Employment Type"),
		mobile: col(headers, "Mobile No"),
	};
	const missing = Object.keys(ix).filter((k) => ix[k] === null);
	if (missing.length) {
		process.stderr.write("Columns not found in the export: " + missing.join(", ") + "\n");
		process.exit(1);
	}

	const needed = { Department: new Set(), Designation: new Set(), "Employment Type": new Set() };
	const outRows = [];
	const rejected = [];
	const seenCodes = new Set();
	const stats = new Map();
	const bump = (k) => stats.set(k, (stats.get(k) || 0) + 1);

	for (const row of rows.slice(index + 1)) {
		const code = get(row, ix.code);
		if (!code) continue;

		const statusRaw = get(row, ix.status).toLowerCase();
		const status = STATUS[statusRaw];
		if (!status) { rejected.push([code, `unrecognised status '${statusRaw}'`]); continue; }

		if (status === "Left" && !values["include-left"]) { bump("skipped_leaver"); continue; }

		if (seenCodes.has(code)) { rejected.push([code, "duplicate employee code"]); continue; }
		seenCodes.add(code);

		const company = COMPANY[get(row, ix.company).toUpperCase()];
		if (!company) {
			rejected.push([code, `company not in the mapping: '${get(row, ix.company)}'`]);
			continue;
		}

		const doj = asDate(get(row, ix.doj));
		if (!doj) {
			/* ERPNext refuses an Employee with no joining date, and a guessed one
			   would quietly change somebody's length of service and gratuity. */
			rejected.push([code, "no usable joining date"]);
			continue;
		}

		const full = get(row, ix.name);
		const [first, last] = splitName(full);
		const gender = GENDER[get(row, ix.gender).toLowerCase()] || "";
		if (!gender) bump("blank_gender");

		const device = get(row, ix.machine);
		if (status === "Active" && !device) bump("active_without_device_id");

		const dept = departmentName(get(row, ix.dept), company);
		const desig = tidy(get(row, ix.desig));
		const etype = get(row, ix.etype);

		outRows.push({
			"ID": "",
			"Employee Name": full,
			"First Name": first,
			"Last Name": last,
			"Gender": gender,
			"Date of Birth": asDate(get(row, ix.dob)),
			"Date of Joining": doj,
			"Relieving Date": asDate(get(row, ix.left)),
			"Status": status,
			"Company": company,
			"Department": dept,
			"Designation": desig,
			"Employment Type": etype,
			"Attendance Device ID (Biometric/RF tag ID)": device,
			"Cell Number": get(row, ix.mobile),
			/* No Holiday List column on purpose. It is set once on each Company
			   as `default_holiday_list`, and Employee falls back to it. Repeating
			   the name on 160 rows means 160 rows to edit next April. */
			"Factor HR ID": code,
		});
		bump("exported");
		bump("company:" + company);
		needed.Department.add(dept);
		needed.Designation.add(desig);
		needed["Employment Type"].add(etype);
	}

	ensureDir(OUT_DIR);
	const outPath = path.join(OUT_DIR, "employee-import.csv");
	writeCsv(outPath, HEADERS, outRows.map((r) => HEADERS.map((h) => r[h])));

	const wb = new ExcelJS.Workbook();
	wb.creator = "Manna HR tools";
	const ws = wb.addWorksheet("Employee");
	ws.addRow(HEADERS);
	for (const r of outRows) ws.addRow(HEADERS.map((h) => r[h]));
	const xlsxPath = path.join(OUT_DIR, "03-employee.xlsx");
	await wb.xlsx.writeFile(xlsxPath);
	console.log(`wrote ${xlsxPath}`);

	console.log(`wrote ${outPath}  (${outRows.length} rows)`);
	console.log();
	for (const key of [...stats.keys()].sort()) {
		console.log(`  ${padEnd(key, 44)} ${stats.get(key)}`);
	}

	/* Every Link value the import will need. A missing one fails that row at
	   import time, one row at a time, which is a slow way to find out. */
	const mastersPath = path.join(OUT_DIR, "masters-needed.csv");
	const masterRows = [];
	for (const dt of Object.keys(needed).sort()) {
		for (const value of [...needed[dt]].filter(Boolean).sort()) masterRows.push([dt, value]);
	}
	writeCsv(mastersPath, ["Doctype", "Value"], masterRows);
	console.log("");
	console.log(`masters the import will need -> ${mastersPath}`);
	for (const dt of Object.keys(needed).sort()) {
		console.log(`  ${padEnd(dt, 18)} ${[...needed[dt]].filter(Boolean).length} distinct`);
	}

	if (rejected.length) {
		const rejPath = path.join(OUT_DIR, "employee-rejected.csv");
		writeCsv(rejPath, ["Emp Code", "Why"], rejected);
		console.log(`\n  *** ${rejected.length} row(s) rejected -> ${rejPath}`);
		for (const [code, why] of rejected.slice(0, 10)) {
			console.log(`      ${padEnd(code, 12)} ${why}`);
		}
	}
}

main();

File: tools/build_employee_import.py
"""Turn the Factor HR employee master into an ERPNext Data Import file.

    python tools/build_employee_import.py                 # active staff only
    python tools/build_employee_import.py --include-left  # everyone, leavers too

Writes `data/out/employee-import.csv`, which is gitignored along with everything
else under `data/`.

Two things this deliberately does **not** do:

  - **It never guesses a company.** Six spellings in Factor HR map onto five
    ERPNext companies through the table below, and a row whose company is not in
    that table is rejected rather than defaulted. A guessed company is a person
    in the wrong payroll, discovered at month end.
  - **It does not set `reports_to`.** A manager has to exist before anyone can
    report to them, so that is a second pass — see `--pass2`.
"""

import argparse
import csv
import os
import re
import sys
from collections import Counter

try:
	from openpyxl import load_workbook
except ImportError:
	sys.exit("Needs openpyxl: python -m pip install openpyxl")

SOURCE = "data/factohr/Employee Detail Report.xlsx"
OUT_DIR = "data/out"

# Factor HR spelling -> ERPNext Company. Confirmed with Manna on 22 Aug 2026.
#
# Note the two rows that collapse onto Manna Treads: HI-TECH PRETREADS is the
# same company, not a separate one, so its 112 people join the 4 already there.
COMPANY = {
	"MANNA RUBBER PRODUCTS PVT.LTD.": "Manna Rubber Products Private Limited",
	"HI-TECH PRETREADS": "Manna Treads",
	"MANNA TREADS PVT.LTD": "Manna Treads",
	"HI-TECH RUBBER INDUSTRIES": "Hi-Tech Rubber Industries",
	"MANNA TYRE RETREADS": "Manna Tyre Retreads",
	"MANNA GROUP H-QTRS": "Manna Group Headquarters",
}

# ERPNext names a Department "<name> - <company abbr>", so a bare "Production"
# links to nothing and the import fails a row at a time. Designation is not
# suffixed. This caught us on the first build.
ABBR = {
	"Manna Rubber Products Private Limited": "MRPPL",
	"Manna Treads": "MT",
	"Manna Tyre Retreads": "MTR",
	"Hi-Tech Rubber Industries": "HRI",
	"Manna Group Headquarters": "MGHQ",
	"Manna Tyre UAE": "MRU",
}

# ERPNext's Employee.status is a Select with exactly these options.
STATUS = {"active": "Active", "inactive": "Left"}

GENDER = {"male": "Male", "female": "Female"}

# Columns written, in the order ERPNext's importer likes to see them.
HEADERS = [
	"ID",
	"Employee Name",
	"First Name",
	"Last Name",
	"Gender",
	"Date of Birth",
	"Date of Joining",
	"Relieving Date",
	"Status",
	"Company",
	"Department",
	"Designation",
	"Employment Type",
	"Attendance Device ID (Biometric/RF tag ID)",
	"Cell Number",
	"Factor HR ID",
]


def find_header(rows, want="Emp Code", limit=15):
	target = want.lower()
	for i, row in enumerate(rows[:limit]):
		for cell in row:
			if cell and str(cell).strip().lower() == target:
				return i
	return None


def col(headers, *names):
	lowered = [h.lower() for h in headers]
	for name in names:
		if name.lower() in lowered:
			return lowered.index(name.lower())
	for name in names:
		n = name.lower()
		for i, h in enumerate(lowered):
			if h.startswith(n):
				return i
	return None


def get(row, i):
	if i is None or i >= len(row):
		return ""
	v = row[i]
	return "" if v is None else str(v).strip()


def as_date(value):
	"""Factor HR emits '1966-12-04 00:00:00'. ERPNext wants '1966-12-04'."""
	if not value:
		return ""
	m = re.match(r"(\d{4}-\d{2}-\d{2})", value)
	if m:
		return m.group(1)
	m = re.match(r"(\d{2})-([A-Za-z]{3})-(\d{4})", value)
	if m:
		months = dict(jan="01", feb="02", mar="03", apr="04", may="05", jun="06",
		              jul="07", aug="08", sep="09", oct="10", nov="11", dec="12")
		mon = months.get(m.group(2).lower())
		if mon:
			return "{0}-{1}-{2}".format(m.group(3), mon, m.group(1))
	return ""


def split_name(full):
	"""ERPNext requires `first_name`; `last_name` is optional.

	Many rows here are a single word, and several are a full name in one field
	with no reliable split. Putting everything in `first_name` and letting
	`employee_name` carry the display form is honest — inventing a surname from
	the last token would corrupt names that do not work that way.
	"""
	full = " ".join(full.split())
	if not full:
		return "", ""
	parts = full.split(" ")
	if len(parts) == 1:
		return parts[0], ""
	return " ".join(parts[:-1]), parts[-1]


def tidy(value):
	"""Factor HR shouts everything. Title-case it, but leave initialisms alone."""
	value = " ".join(value.split())
	if not value:
		return ""
	out = []
	for word in value.split(" "):
		# Q.C., H.R, PVT — anything with a dot or 3 letters or fewer that is all
		# caps is almost certainly an initialism, not a shouted word.
		if "." in word or (word.isupper() and len(word) <= 3):
			out.append(word)
		else:
			out.append(word.capitalize())
	return " ".join(out)


def department_name(raw, company):
	if not raw:
		return ""
	return "{0} - {1}".format(tidy(raw), ABBR[company])


def main():
	ap = argparse.ArgumentParser()
	ap.add_argument("--source", default=SOURCE)
	ap.add_argument("--include-left", action="store_true",
	                help="also export the 344 leavers, as Status=Left")
	args = ap.parse_args()

	book = load_workbook(args.source, read_only=True, data_only=True)
	ws = book.worksheets[0]
	rows = [list(r) for r in ws.iter_rows(max_row=3000, values_only=True)]
	book.close()

	index = find_header(rows)
	if index is None:
		sys.exit("Could not find an 'Emp Code' column in " + args.source)
	headers = [str(c).strip() if c is not None else "" for c in rows[index]]

	ix = {
		"code": col(headers, "Emp Code"),
		"name": col(headers, "Full Name"),
		"company": col(headers, "Company Name"),
		"status": col(headers, "Status"),
		"machine": col(headers, "Machine Code"),
		"dob": col(headers, "Birth Date"),
		"doj": col(headers, "Joining Date"),
		"left": col(headers, "Leaving Date"),
		"gender": col(headers, "Gender"),
		"dept": col(headers, "Department"),
		"desig": col(headers, "Designation"),
		"etype": col(headers, "Employment Type"),
		"mobile": col(headers, "Mobile No"),
	}
	missing = [k for k, v in ix.items() if v is None]
	if missing:
		sys.exit("Columns not found in the export: " + ", ".join(missing))

	from collections import defaultdict
	needed = defaultdict(set)
	out_rows, rejected = [], []
	seen_codes = set()
	stats = Counter()

	for row in rows[index + 1:]:
		code = get(row, ix["code"])
		if not code:
			continue

		status_raw = get(row, ix["status"]).lower()
		status = STATUS.get(status_raw)
		if not status:
			rejected.append((code, "unrecognised status '{0}'".format(status_raw)))
			continue

		if status == "Left" and not args.include_left:
			stats["skipped_leaver"] += 1
			continue

		if code in seen_codes:
			rejected.append((code, "duplicate employee code"))
			continue
		seen_codes.add(code)

		company = COMPANY.get(get(row, ix["company"]).upper())
		if not company:
			rejected.append((code, "company not in the mapping: '{0}'".format(get(row, ix["company"]))))
			continue

		doj = as_date(get(row, ix["doj"]))
		if not doj:
			# ERPNext refuses an Employee with no joining date, and a guessed one
			# would quietly change somebody's length of service and gratuity.
			rejected.append((code, "no usable joining date"))
			continue

		full = get(row, ix["name"])
		first, last = split_name(full)
		gender = GENDER.get(get(row, ix["gender"]).lower(), "")
		if not gender:
			stats["blank_gender"] += 1

		device = get(row, ix["machine"])
		if status == "Active" and not device:
			stats["active_without_device_id"] += 1

		out_rows.append({
			"ID": "",
			"Employee Name": full,
			"First Name": first,
			"Last Name": last,
			"Gender": gender,
			"Date of Birth": as_date(get(row, ix["dob"])),
			"Date of Joining": doj,
			"Relieving Date": as_date(get(row, ix["left"])),
			"Status": status,
			"Company": company,
			"Department": department_name(get(row, ix["dept"]), company),
			"Designation": tidy(get(row, ix["desig"])),
			"Employment Type": get(row, ix["etype"]),
			"Attendance Device ID (Biometric/RF tag ID)": device,
			"Cell Number": get(row, ix["mobile"]),
			# No Holiday List column on purpose. It is set once on each Company
			# as `default_holiday_list`, and Employee falls back to it. Repeating
			# the name on 160 rows means 160 rows to edit next April.
			"Factor HR ID": code,
		})
		stats["exported"] += 1
		stats["company:" + company] += 1
		needed["Department"].add(department_name(get(row, ix["dept"]), company))
		needed["Designation"].add(tidy(get(row, ix["desig"])))
		needed["Employment Type"].add(get(row, ix["etype"]))

	os.makedirs(OUT_DIR, exist_ok=True)
	out_path = os.path.join(OUT_DIR, "employee-import.csv")
	with open(out_path, "w", newline="", encoding="utf-8-sig") as fh:
		w = csv.DictWriter(fh, fieldnames=HEADERS)
		w.writeheader()
		w.writerows(out_rows)

	try:
		from openpyxl import Workbook
		wb = Workbook(); ws = wb.active; ws.title = "Employee"
		ws.append(HEADERS)
		for r in out_rows:
			ws.append([r[h] for h in HEADERS])
		xlsx_path = os.path.join(OUT_DIR, "03-employee.xlsx")
		wb.save(xlsx_path)
		print("wrote {0}".format(xlsx_path))
	except ImportError:
		pass

	print("wrote {0}  ({1} rows)".format(out_path, len(out_rows)))
	print()
	for key in sorted(stats):
		print("  {0:44} {1}".format(key, stats[key]))

	# Every Link value the import will need. A missing one fails that row at
	# import time, one row at a time, which is a slow way to find out.
	masters_path = os.path.join(OUT_DIR, "masters-needed.csv")
	with open(masters_path, "w", newline="", encoding="utf-8-sig") as fh:
		w = csv.writer(fh)
		w.writerow(["Doctype", "Value"])
		for dt in sorted(needed):
			for value in sorted(v for v in needed[dt] if v):
				w.writerow([dt, value])
	print("")
	print("masters the import will need -> {0}".format(masters_path))
	for dt in sorted(needed):
		print("  {0:18} {1} distinct".format(dt, len([v for v in needed[dt] if v])))

	if rejected:
		rej_path = os.path.join(OUT_DIR, "employee-rejected.csv")
		with open(rej_path, "w", newline="", encoding="utf-8-sig") as fh:
			w = csv.writer(fh)
			w.writerow(["Emp Code", "Why"])
			w.writerows(rejected)
		print("\n  *** {0} row(s) rejected -> {1}".format(len(rejected), rej_path))
		for code, why in rejected[:10]:
			print("      {0:12} {1}".format(code, why))


if __name__ == "__main__":
	main()
